### DISCUSSION TITLE
Raylib 4.5 - Framerate somehow limited by monitor refresh rate

### DISCUSSION THREAD

**Discussion by therion-forever** [raysan5/raylib#3041] [Q&A] [answered]
No matter what I do, the FPS count will be no higher tah my monitor refresh rate: 75Hz

Using 4.5 in C (compiling using N++, just the way it was downloaded this week, under Win10 (64) in a laptop with i7 (6gen) and nvidia 980m.

Even not usin SetTargetFPS() or using a high value like 240 or whatever, I´ll get no more than 75 fps (even drawing a single moving circle)... and yes, when I draw 10k circles, it drops to like 10 fps... same speed I get with p5JS... so the fps seems right. With few balls, p5JS shows like 300fps, but raylib will nott go above 75, my monitor rate.

The same result, measuring using getFPS(), calculation delta time or using GetFrameTime()... all 3 options I get the same values.

Is it something my hardware? Vsync?? any trick I need to do to free it, like SetTargetFPS(-1)??
It won´t happen with p5JS or Pygame.... but it also happens with Processing (qhich I know is by design).

Another point:

Using the exact same algorithm (very simple one, where I create arrays por random positions,radius, colors and velocities for N circles, draw de circles, apply velicity to change position, and check collision bith borders... child easy testing stuff,  I get same performance for both C Raylib (desktop) and p5JS, BUT I get twice the performance (for any N, from 100 to 20000, does not matter) using Pygame2. I believe Pygame2 is using GPU, not old SDL anymore... also, I´m using standard interpreter, not Pypy, not compiling with Nuitka, not using jit through numba, not using numpy for the arrays, NOTHING... simply the very same translated simple code... and draws twice as fast. I was expecting C Raylib to be like 4x faster, not 0.5x

I´ll try to paste the code later here, it´s local in another pc.

I´ll try SetCofigFlags(FLAG_VSYNC_HINT) to disabe Vsync and see if it helps for my hw.

Thank you guys!!!!

**Comment by chriscamacho** [comment]
meeting people half way with info on your hardware and OS will likely get you more help in future

this issue is more complex than many people appreciate and the more information you can give the better.

**Reply by therion-forever** [reply]
Thanks for the reply... that´s what I was looking for, to know if it´s an easy trick or if it´s a complex issue. So it seems to be complex.

As I said, hw is laptop with 6gen i7 and 980m nvidia gpu, Win10 Pro 64bit.
Have just turned VSYNC off, but no changes  -> SetCofigFlags(FLAG_VSYNC_HINT)

Regarding the performance:
I tried 2 different ways: Using sprites (textures) and using shapes (drawing filled circles), using both Pygame and C Raylib. (Standard Python 3.7 interpreter, all with same algorithm (no sprite rotation, just borders collision), no numpy, no jit, nothing). As a bonus, also used p5JS for shapes.

First using circle shape:

Pygame reached 60 fps for 7.5K circles
C Raylib reached 60fps for 1.7K circles
p5JS reached 60 fps for 1.7K circles   

So, Pygame, for circle shape drawing, was 4x faster than C Raylib or p5JS.

Then I changed for sprite (40x40 png, no rotation):
Pygame reached 60 fps for 3K sprites (performance dropped 50% from shapes to sprite, yes, I used convert_alpha())
C Raylib reached 60 fps with 95K sprites
p5JS not tested.

So, Pygame got half of previous performance and C Raylib improved 53x, being 31x faster that Pygame using textures.

For some reason, textures are easy for C Raylib and a pain for Python... and shapes are a bit easy for Pygame and a real pain for C Raylib.


BUT I'm still stuck at 75 FPS (also tried DrawFPS(), got the same.

Here is the code;

/////////////////////////////////////////////

#include "raylib.h"
#include <stdlib.h>  

    typedef struct Ball {
    Vector2 position;
    Vector2 speed;
    Color color;
    int radius;
} Ball;

int main()
{
    const int W= 1600;
    const int H = 800;
    
    //SetConfigFlags(FLAG_VSYNC_HINT);
    
    InitWindow(W, H, "Sander");
	
    SetTargetFPS(240);
    
    const int N = 120000;
    
    Texture2D textura = LoadTexture("ball2.png");

    Ball *balls = (Ball *)malloc(N*sizeof(Ball));  
    
    for (int i = 0; i < N; i++)
    {
                    balls[i].position.x = (int)GetRandomValue(50, W-50);
                    balls[i].position.y = (int)GetRandomValue(50, H-50);
                    balls[i].speed.x = (float)GetRandomValue(-4.0, 4.0);
                    balls[i].speed.y = (float)GetRandomValue(-4.0, 4.0);
                    balls[i].color = (Color){ GetRandomValue(50, 240), GetRandomValue(80, 240), GetRandomValue(100, 240), 255 };
                    balls[i].radius = (int)GetRandomValue(5, 10);
    }
	
	while(!WindowShouldClose())
	{
        BeginDrawing();
		
	ClearBackground(RAYWHITE);
        
        for (int i = 0; i < N; i++)
        {
            balls[i].position.x += balls[i].speed.x;
            balls[i].position.y += balls[i].speed.y;
            
            if (((balls[i].position.x + 30) > W) ||
                ((balls[i].position.x + 30) < 0)) balls[i].speed.x *= -1;
            if (((balls[i].position.y + 30) > H) ||
                ((balls[i].position.y + 30) < 0)) balls[i].speed.y *= -1;
            
            //DrawCircle((int)balls[i].position.x, (int)balls[i].position.y, (int)balls[i].radius, balls[i].color);
            DrawTexture(textura, (int)balls[i].position.x, (int)balls[i].position.y, balls[i].color);
		}
        
        DrawFPS(50,50);

	EndDrawing();

	}
	
    free(balls); 

    UnloadTexture(textura);  

    CloseWindow();
	
    return 0;
}

**Comment by chriscamacho** [comment, marked as answer]
ah! be careful with the 2d stuff, some of it isn't accelerated...

sorry i missed your spec in the first post, not a windows user (except at work *sigh*) but I'd suggest digging around in the gpu settings...

**Reply by therion-forever** [reply]
I´ll take a look at it... also I´ll try to execute it in another machine (first I´ll take only the .exe because I don´t know if I´ll actually have to recompile it in the other machine to see any changes).

For the performance issue, after changing C raylib to texture drawing instead of shape, I did the same with Pygame, using the same small png.... and while in C Raylib it brought great improvement, in Pygame it drove way worst results...

So it seems Pygame is faster thay C Raylib to draw shapes, but C Raylib is much much faster than Pygame to draw sprites... the performance of C Raylib using sprites is really fantastic compared to shapes (I just tested circles, that probably has a high cost to be drawn, and we can see that when drawing thousands of them).

**Reply by therion-forever** [reply]
> ah! be careful with the 2d stuff, some of it isn't accelerated...
> 
> sorry i missed your spec in the first post, not a windows user (except at work _sigh_) but I'd suggest digging around in the gpu settings...

I've just turned G-Sync off in NVIDIA Control panel, and now C Raylib is free to go above my monitor refresh rate! Thanks!

**Comment by therion-forever** [comment]
I've just turned G-Sync off in NVIDIA Control panel, and now C Raylib is free to go above my monitor refresh rate!

**Reply by chriscamacho** [reply]
glad you got it sorted
